gusto/transport_equation.py

```diff
@@ -0,0 +1,345 @@
+from __future__ import absolute_import
+from abc import ABCMeta, abstractmethod
+from firedrake import Function, TestFunction, TrialFunction, \
+    FacetNormal, \
+    dx, dot, grad, div, jump, avg, dS, dS_v, dS_h, inner, \
+    outer, sign, cross, CellNormal, as_vector, sqrt, Constant, \
+    curl, BrokenElement, FunctionSpace, Projector
+
+
+class TransportEquation(object):
+    """
+    Base class for transport equations in Gusto.
+
+    The equation is assumed to be in the form:
+
+    q_t + L(q) = 0
+
+    where q is the (scalar or vector) field to be solved for.
+
+    :arg state: :class:`.State` object.
+    :arg V: :class:`.FunctionSpace object. The function space that q lives in.
+    :arg ibp: string, stands for 'integrate by parts' and can take the value
+              None, "once" or "twice". Defaults to "once".
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, state, V, ibp="once"):
+        self.state = state
+        self.V = V
+        self.ibp = ibp
+
+        # set up functions required for forms
+        self.ubar = Function(state.V[0])
+        self.test = TestFunction(V)
+        self.trial = TrialFunction(V)
+
+        # find out if we are CG
+        nvertex = V.ufl_domain().ufl_cell().num_vertices()
+        entity_dofs = V.fiat_element.entity_dofs()
+        # If there are as many dofs on vertices as there are vertices,
+        # assume a continuous space.
+        try:
+            self.is_cg = sum(map(len, entity_dofs[0].values())) == nvertex
+        except KeyError:
+            self.is_cg = sum(map(len, entity_dofs[(0, 0)].values())) == nvertex
+
+        # DG, embedded DG and hybrid SUPG methods need surface measures,
+        # n and un
+        if self.is_cg:
+            self.dS = None
+        else:
+            if V.extruded:
+                self.dS = (dS_h + dS_v)
+            else:
+                self.dS = dS
+            self.n = FacetNormal(state.mesh)
+            self.un = 0.5*(dot(self.ubar, self.n) + abs(dot(self.ubar, self.n)))
+
+    def mass_term(self, q):
+        return inner(self.test, q)*dx
+
+    @abstractmethod
+    def advection_term(self):
+        pass
+
+
+class LinearAdvection(TransportEquation):
+    """
+    Class for linear transport equation.
+
+    :arg state: :class:`.State` object.
+    :arg V: :class:`.FunctionSpace object. The function space that q lives in.
+    :arg qbar: The reference function that the equation has been linearised
+               about. It is assumed that the reference velocity is zero and
+               the ubar below is the nonlinear advecting velocity
+               0.5*(u'^(n+1) + u'(n)))
+    :arg ibp: string, stands for 'integrate by parts' and can take the value
+              None, "once" or "twice". Defaults to "once".
+    :arg continuity: is the equation in continuity form L(q) = div(u'*qbar)?
+                     If False, the equation is assumed to be in advective
+                     form L(q) = u' dot grad(qbar)
```

Comment:
For ease of readability, I'd be tempted to change this flag into an argument:  `equation_form` which takes the value "continuity" or "advective".